### ISSUE TITLE
Printer listings take a long time to refresh on CUPS

### ISSUE THREAD

**Issue by tresf** [AdoptOpenJDK/openjdk-build#1212]
**Platform:**
MacOS, Linux (Issue does NOT occur on Windows)

**Architecture:**
x86_64 (AdoptOpenJDK 11.0.4.  Also occurs on Oracle Java 8).

**Description**
Java's `PrintService.lookupPrintServices()` does not return newly added (and thus still returns newly removed) printers in a timely fashion unless the Java application is restarted.

**Steps**

1. Run the following code:
    ```java
    import javax.print.PrintService;
    import javax.print.PrintServiceLookup;
    
    public class ListingDelay {
    
        public static void main(String[] args) {
            new Thread(() -> {
                try {
                    while(true) {
                        PrintService[] printers = PrintServiceLookup.lookupPrintServices(null, null);
                        System.out.println("Found " + printers.length + " printers.  (Sleeping for 5 seconds)");
                        Thread.sleep(5000);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }).run();
        }
    }
    ```
2. Add or remove a printer.  The number will not change for 1-2 minutes.
3. Restart the program.  Now the number will change.

**Comment by karianna** [comment]
Can confirm that this works correctly (as per the code) for Ubuntu's 11.0.4 but not for Adopt's

**Comment by tresf** [comment]
@karianna can you share some details with your unit tests?  They do not match mine.

I've tested the above code using Ubuntu 18.04 and the openjdk version from Canonical and the issue exists there for me as well.

```
/usr/lib/jvm/java-11-openjdk-amd64/bin/java -version
openjdk version "11.0.4" 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)
```

I installed a local Zebra printer and the CUPS-PDF virtual printer:

Start up the code snippet.
* Java correctly reports 2 printers.
* Adding an additional printer, java incorrectly reports 2 printers (should be 3)

Start up the code snippet.
* Java correctly reports 3 printers.
* Removing a printer, java incorrectly reports 3 printers (should be 2)

**Comment by karianna** [comment]
> > this works correctly or Ubuntu's 11.0.4
> 
> @karianna can you share some details with your unit tests? They do not match mine.
> 
> I've tested the above code using Ubuntu 18.04 and the openjdk version from Canonical and the issue exists there for me as well.
> 
> ```
> /usr/lib/jvm/java-11-openjdk-amd64/bin/java -version
> openjdk version "11.0.4" 2019-07-16
> OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
> OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)
> ```
> 
> I installed a local Zebra printer and the CUPS-PDF virtual printer:
> 
> Start up the code snippet.
> 
> * Java correctly reports 2 printers.
> * Adding an additional printer, java incorrectly reports 2 printers (should be 3)
> 
> Start up the code snippet.
> 
> * Java correctly reports 3 printers.
> * Removing a printer, java incorrectly reports 3 printers (should be 2)

I'm using 18.04.3 (LTS) and canonical's 11.0.4

```
openjdk 11.0.4 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)
```

I run your test above and when adding and removing the CUPS-PDF virtual printer it adds and takes away the number as expected.  With Adopt 11.0.4, it does not.

One can only assume we're referencing a different version of a native library.

**Comment by tresf** [comment]
> One can only assume we're referencing a different version of a native library.

Hmm... I'll update the machine and test again, but that just makes the symptoms even more suspicious.

**Comment by tresf** [comment]
Machine updated.  I can confirm that openjdk11 from Ubuntu refreshes the printer list although it takes 1-2 minutes to do so.

I also have the same results on Adopt, 1-2 minutes the printer listing updates.  Can you confirm that waiting up to 2 minutes, it finally updates?

**Comment by tresf** [comment]
Same results for MacOS.... waiting about 2 minutes and the printer listing refreshes.

![image](https://user-images.githubusercontent.com/6345473/62958755-614aa680-bdc5-11e9-8819-573521aef8b5.png)

**Comment by tresf** [comment]
Updated bug report to be more descriptive.

**Comment by karianna** [comment]
Thanks for the investigation and updates!  I'll run mine again...

When I add a printer it picks that up in 10-15 seconds, removing took about 1-2 minutes.  I suspect this is falling within 30-60 second caching / timing windows with the O/S.

Aha - try adding `-Dsun.java2d.print.polling=false` to the java command line before executing your program.

e.g. `java -Dsun.java2d.print.polling=false ListingDelay`

That seems to add and remove the printers roughly at the speed of them being added/removed from the O/S config.

**Comment by tresf** [comment]
@karianna that works very well.

Specifically:

```java
System.setProperty("sun.java2d.print.polling", "false");
```

This adds about `90ms` of lookup time to the `PrintServiceLookup.lookupPrintServices(...);` call (with polling on, Java measures this as `0ms`), but all previous attempts to force-refresh (by hacking the awt hash) added `200ms` - `200,000ms`.

This seems like a perfectly viable workaround.  Do you know if this affects the Windows lookup?

**Comment by karianna** [comment]
> @karianna that works very well.
> 
> Specifically:
> 
> ```java
> System.setProperty("sun.java2d.print.polling", "false");
> ```
> 
> This adds about `90ms` of lookup time to the `PrintServiceLookup.lookupPrintServices(...);` call (with polling on, Java measures this as `0ms`), but all previous attempts to force-refresh (by hacking the awt hash) added `200ms` - `200,000ms`.
> 
> This seems like a perfectly viable workaround. Do you know if this affects the Windows lookup?

Yeah, it seems really counterintuitive that setting that variable to false would speed things up.  It had the same effect on Windows for me, so based on that small sample set I think you can roll out the workaround.  I'll get a bug filed upstream as well.

**Comment by tresf** [comment]
FYI, I ran some tests and after about 100,000 calls to `PrintServiceLookup.lookupPrintServices`, the delay increases (gradually) to about 1,000 ms.  On my machine (i7, Quad Core) 100,000 calls took about a day.

This still outperforms any custom refresh techniques that I'm using today, but help explains why it's true by default.  In comparison, with this setting turned off, 100,000 calls to `PrintServiceLookup.lookupPrintServices` occurs in less than one second because there's no delay between calls.

**Comment by tresf** [comment]
Another slight downside to this technique is once `PrintServiceLookup.lookupPrintServices` has been called, the behavior can never be toggled back off.  There's no way to get the `PrintServiceLookup.lookupPrintServices` back to the super fast 0ms response time that it takes naturally.  If the first call to `PrintServiceLookup.lookupPrintServices(...)` happens when `sun.java2d.print.polling` parameter is `false`, it stays that way forever (`80ms` between calls, gradually increasing to `1,000ms`)

**Comment by tresf** [comment]
> It had the same effect on Windows for me

@karianna Per original bug report, I do not experience the underlying issue on Windows at all.  Can you be more specific?

**Comment by karianna** [comment]
I also did not have the bug on windows but adding that property setting had no ill effects (in fact behavior didn't see to change at all)

**Comment by tresf** [comment]
@karianna that makes more sense as Windows seems to have resolved these issues: https://bugs.openjdk.java.net/browse/JDK-8153732

**Comment by karianna** [comment]
Oh nice find - I'll reference that issue when we open a new one 
https://bugs.openjdk.java.net/browse/JDK-8153732

**Comment by jerboaa** [comment]
Reported upstream as: https://bugs.openjdk.java.net/browse/JDK-8229869

**Comment by tresf** [comment]
@karianna this is in closed status after filing upstream, but I'm not sure how I would be able to monitor the status of upstream. JDK's bug system says the following:

> Everyone with OpenJDK Author status or above has a JBS account which may be used to create and edit bugs. Those without accounts can view bugs anonymously.

Is the expectation for the OPs to constantly monitor upstream manually?

**Comment by karianna** [comment]
I swear there did use to be an RSS feed out of those bugs but I'll be darned if I can't find it.  Reopening this in that case as we at least sweep these from time to time.
